src/libraries/System.Net.Mail/src/System/Net/Mail/SmtpNegotiateAuthenticationModule.cs

```diff
@@ -1,92 +1,84 @@
 
+using System.Buffers;
 using System.Collections.Generic;
 using System.ComponentModel;
+using System.Net.Security;
 using System.Security.Authentication.ExtendedProtection;
 
 namespace System.Net.Mail
 {
     internal sealed class SmtpNegotiateAuthenticationModule : ISmtpAuthenticationModule
     {
         private static byte[] _saslNoSecurtyLayerToken = new byte[] { 1, 0, 0, 0 };
-        private readonly Dictionary<object, NTAuthentication> _sessions = new Dictionary<object, NTAuthentication>();
+        private readonly Dictionary<object, NegotiateAuthentication> _sessions = new Dictionary<object, NegotiateAuthentication>();
 
         internal SmtpNegotiateAuthenticationModule()
         {
         }
 
         public Authorization? Authenticate(string? challenge, NetworkCredential? credential, object sessionCookie, string? spn, ChannelBinding? channelBindingToken)
         {
-            try
+            lock (_sessions)
             {
-                lock (_sessions)
+                NegotiateAuthentication? clientContext;
+                if (!_sessions.TryGetValue(sessionCookie, out clientContext))
                 {
-                    NTAuthentication? clientContext;
-                    if (!_sessions.TryGetValue(sessionCookie, out clientContext))
+                    if (credential == null)
                     {
-                        if (credential == null)
-                        {
-                            return null;
-                        }
-
-                        ContextFlagsPal contextFlags = ContextFlagsPal.Connection | ContextFlagsPal.InitIntegrity;
-                        // Workaround for https://github.com/gssapi/gss-ntlmssp/issues/77
-                        // GSSAPI NTLM SSP does not support gss_wrap/gss_unwrap unless confidentiality
-                        // is negotiated.
-                        if (OperatingSystem.IsLinux())
-                        {
-                            contextFlags |= ContextFlagsPal.Confidentiality;
-                        }
-
-                        _sessions[sessionCookie] =
-                            clientContext =
-                            new NTAuthentication(false, "Negotiate", credential, spn,
-                                                 contextFlags, channelBindingToken);
+                        return null;
                     }
 
-                    byte[]? byteResp;
-                    string? resp = null;
-
-                    if (!clientContext.IsCompleted)
+                    ProtectionLevel protectionLevel = ProtectionLevel.Sign;
+                    // Workaround for https://github.com/gssapi/gss-ntlmssp/issues/77
+                    // GSSAPI NTLM SSP does not support gss_wrap/gss_unwrap unless confidentiality
+                    // is negotiated.
+                    if (OperatingSystem.IsLinux())
                     {
+                        protectionLevel = ProtectionLevel.EncryptAndSign;
```

Comment:
Is there harm of doing it also on WindowsOS/mac?

Reply:
(FWIW: this is not a new check in this PR, it is just re-indented existing one)

In theory it would be fine on all systems but I don't have sufficient data to confirm. The negotiated protocol can end up being either Kerberos or NTLM. If it's Kerberos on Linux/macOS we get forced confidentiality anyway. For all other scenarios it could result in unilaterally bumped security requirements from the client side. If the server is Exchange then it would work. If the server is something running on an embedded device with Cyrus SASL backend I would not be so sure that everything would still work (notably the NTLM implementation in Cyrus is not even interoperable with Windows 11 in default configuration).

Reply:
Short version: It's mostly just being cautious.

Reply:
That is fine. I was nice to see less platform specific code through this PR and if we could avoid one more platform check I would be happy. But even with it it seems like great improvemet.